src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -104,3 +118,233 @@ def snapshot(self) -> Iterable[JobInfo]:
+    task: RunTask, job_run: BaseRun, all_clusters: dict[str, ClusterDetails]
+) -> compute.ClusterSpec | ClusterDetails | None:
+    """
+    Determine the cluster associated with the task
+    1) Look for new_cluster on the task
+    2) Look for existing cluster on the task
+    Note - this function has DIFFERENT logic from the _get_cluster_configs_from_all_jobs
+    It iterates over the tasks of a submit run, and not the job clusters of a job
+    """
+    if task.new_cluster is not None:
+        return task.new_cluster
+    if task.existing_cluster_id is not None:
+        # from api docs If existing_cluster_id, the ID of an existing cluster that is used for all runs of this job.
+        if job_run.job_clusters is not None:
+            job_clusters: dict[str, JobCluster] = {x.job_cluster_key: x for x in job_run.job_clusters}
+            if job_clusters.get(task.existing_cluster_id, None) is not None:
+                cluster_optional = job_clusters.get(task.existing_cluster_id)
+                if cluster_optional is not None:
+                    return cluster_optional.new_cluster
+    if task.cluster_instance is not None and task.cluster_instance.cluster_id is not None:
+        return all_clusters.get(task.cluster_instance.cluster_id)
+    return None
+
+
+def is_custom_image(version_string: str) -> bool:
+    return "custom" in version_string.lower()
+
+
+def spark_version_greater_than(left_version: str | None, right_version: str | None) -> bool:
+    """
+    Is left version greater than right version
+    """
+    if left_version is None or is_custom_image(left_version):
+        return False
+    if right_version is None or is_custom_image(right_version):
+        return True
+    pattern = r"^(?P<major>\d+)?\.(?P<minor>\d+)?\.(?P<patch>[\dx]+)?.*"
+    lvg = re.match(pattern, left_version)
+    rvg = re.match(pattern, right_version)
+    if lvg is None or rvg is None:
+        return False
+    left = (int(lvg.group("major")), int(lvg.group("minor")))
+    right = (int(rvg.group("major")), int(rvg.group("minor")))
+    return left >= right
+
+
+@dataclass
+class SubmitRunInfo:
+    run_id: int | None
+    hashed_id: str  # a pseudo id that combines all the hashable attributes of the run
+    spark_version: str
+    num_tasks_with_failing_configuration: int
+
+
+class SubmitRunsCrawler(CrawlerBase[SubmitRunInfo], JobsMixin):
+
+    def __init__(self, ws: WorkspaceClient, sbe: SqlBackend, schema):
+        super().__init__(sbe, "hive_metastore", schema, "submit_runs", SubmitRunInfo)
+        self._ws = ws
+
+    def snapshot(self) -> Iterable[SubmitRunInfo]:
+        return self._snapshot(self._try_fetch, self._crawl)
+
+    def _submit_runs_by_hour(self, start_date: datetime, end_date: datetime) -> Iterable[BaseRun]:
+        for start, end in generate_hourly_ranges(start_date, end_date):
+            logger.info(f"Fetching submit runs from {start} to {end}")
+            yield from self._ws.jobs.list_runs(
```

Comment:
Why do we need many ranges, I don't understand? SDK does pagination under the hood. By adding hourly ranges you make the code unreadable

Reply:
listing the job runs takes a lot of time (10+ minutes for 1k+ runs). Without a convenient log it's quite unclear for a user whenever there is any progress or not.